content/guides/server-sent-events.adoc

```diff
@@ -10,26 +10,164 @@ toc::[]
+sample applications or the first few link:hello-world[Hello World]
+guides. In particular, you should know how to run a build, start a
+REPL, and start and stop your server.
+
+You do not need to know any other Clojure web frameworks, but if you
+do, you may find some of the comparisons useful. If not, that's OK,
+just skip those sections.
+
 == Getting Help if You're Stuck
 
+If you get stuck at any point in this guide, please submit an
+https://github.com/pedestal/docs/issues[issue] about this guide or hop
+over to the
+https://groups.google.com/forum/#!forum/pedestal-users[mailing list]
+and raise your hand there. You can often find help in the "pedestal"
+channel of the https://clojurians.slack.com[Clojurians Slack team.]
+
 == Where We Are Going
 
+We'll be creating a new Pedestal service project, adding a simple route to stream a counter using server-sent events, and add a simple HTML page that uses a JavaScript EventSource to open a connection to the stream.
+
 == Before We Begin
 
-== Content Section 1
+This guide will use http://leiningen.org/[Leiningen] to generate and
+build a template project. If you haven't already installed it, please
+take a few minutes to set it up.
+
+== Generate Our New Project
+
+----
+$ lein new pedestal-service myapp
+Generating a pedestal-service application called server-sent-events.
+$ cd server-sent-events
+----
+
+== Add Hiccup
+
+We will be using Hiccup to generate our index page, so we will need to add the dependency in our project.clj.
+
+[source,clojure]
+.project.clj
+----
+include::server-sent-events/project.clj[tags=add-hiccup]
+----
+
+== Clean Up
+
+Here's the service.clj file generated for us:
+
+[source,clojure]
+.src/server_sent_events/service.clj
+----
+include::server-sent-events/src/server_sent_events/service.clj[tags=original-service]
+----
+
+Let's simply things by uncommenting the 'Terse/Vector-based routes' section and removing uneccessary code.
+
+[source,clojure]
+.src/server_sent_events/service.clj
+----
+include::server-sent-events/src/server_sent_events/service.clj[tags=cleaned-up]
+----
+
+== Server-Sent Event Route
+
+Now let's add our server-sent event route. First Lets require SSE and core.async.
+
+[source,clojure]
+.src/server_sent_events/service.clj
+----
+include::server-sent-events/src/server_sent_events/service.clj[tags=require-sse]
+----
+
+Now we'll add our counter route. We create the "/counter" route by calling sse/start-event-stream and passing in our stream-ready handler. This creates an interceptor that will handle setting up the connection and call the stream-ready handler. A stream-ready handler receives a core.async event-chan and SSE context as parameters. You send events on the channel using regular core.async (such as async/>!!) functions and terminate the connection by calling async/close! on the channel. Here our event consists the :name and :data keys, but could also contain an :id for the event.
```

Comment:
Here is one of the tricky parts.

The events that you send on the channel can happen any time later. They don't all have to be created and sent in the stream-ready handler. 

But, the stream-ready handler *must* send at least one event in order to properly initialize the stream. If it doesn't, the client will think the stream is broken and keep reconnecting.